broadlink/climate.py

```diff
@@ -201,3 +201,252 @@ def set_schedule(self, weekday: List[dict], weekend: List[dict]) -> None:
+        """Returns a dictionary with the unit's parameters.
+        
+        Args:
+            payload (Optional[bool]): add the received payload for debugging
+        
+        Returns:
+            dict:
+                state (bool): power
+                set_temp (float): temperature set point 16<n<32
+                mode (str): Cooling, Heating, Fan, Dry, Auto (use first letter)
+                speed (str): Mute, Low, Mid, High, Turbo (available only in cooling) (first letter and Mu for Mute)
+                swing_h (str): ON, OFF
+                swing_v (str): ON, OFF, 1, 2, 3, 4, 5 (fixed positions)
+                sleep (bool): 
+                display (bool):
+                health (bool):
+        """
+        payload = self._send_short_payload(1)
+        assert(len(payload) == 32)
+        data = {}
+        data['state'] = True if (payload[0x14] & 0x20) == 0x20 else False
+        data['set_temp'] = (payload[0x0c] >> 3) + 8 + (0.0 if ((payload[0xe] & 0b10000000) == 0) else 0.5)
+
+        swing_v = payload[0x0c] & 0b111
+        swing_h = (payload[0x0d] & 0b11100000) >> 5
+        if (swing_h == 0b111):
+            data['swing_h'] = 'OFF'
+        elif (swing_h == 0b111):
+            data['swing_h'] = 'OFF'
+        else:
+            data['swing_h'] = 'unrecognized value'
+        
+        if (swing_v == 0b111):
+            data['swing_v'] = 'OFF'
+        elif (swing_v == 0b000):
+            data['swing_v'] = 'ON'
+        elif (swing_v >= 0 and swing_v <=5):
+            data['swing_v'] = str(swing_v)
+        else:
+            data['swing_v'] = 'unrecognized value'
+
+        mode = payload[0x11] >> 3 << 3
+        if mode == 0x00:
+            data['mode'] = 'A'
+        elif mode == 0x20:
+            data['mode'] = 'C'
+        elif mode == 0x40:
+            data['mode'] = 'D'
+        elif mode == 0x80:
+            data['mode'] = 'H'
+        elif mode == 0xc0:
+            data['mode'] = 'F'
+        else:
+            data['mode'] = 'unrecognized value'
+
+        speed_L = payload[0x0f]
+        speed_R = payload[0x10] 
+        if speed_L == 0x60 and speed_R == 0x00:
+            data['speed'] = 'L'
+        elif speed_L == 0x40 and speed_R == 0x00:
+            data['speed'] = 'M'
+        elif speed_L == 0x20 and speed_R == 0x00:
+            data['speed'] = 'H'
+        elif speed_L == 0x40 and speed_R == 0x80:
+            data['speed'] = 'Mu'
+        elif speed_R == 0x40:
+            data['speed'] = 'T'
+        elif speed_L == 0xa0 and speed_R == 0x00:
+            data['speed'] = 'A'
+        else:
+            data['speed'] = 'unrecognized value'
+
+        data['sleep'] = True if (payload[0x1a] & 0b100 == 0b000) else False
+        data['display'] = (payload[0x16] & 0x10 == 0x10)
+        data['health'] = (payload[0x14] & 0b11 == 0b11)
+
+        checksum = self._calculate_checksum(payload[:0x19]) # checksum=(payload[0x1a] << 8) + payload[0x19]
```

Comment:
How about something like this?
```python3
        checksum = resp[0x19] | (resp[0x1a] << 8)
        if sum(resp[:0x19], 0x20017) - sum(resp[0x19:0x1b]) & 0xffff != checksum:
            raise exception(-4008)  # Checksum error.
```
Are you sure 0x20017 is not an int (20017)? Please double check, this number does not feel right.

If so, you can use this:
```python3
        checksum = resp[0x19] | (resp[0x1a] << 8)
        if sum(resp[:0x19], 0x4e31) - sum(resp[0x19:0x1b]) & 0xffff != checksum:
            raise exception(-4008)  # Checksum error.
```

Reply:
I don't fully understand how the checksum works, but I've done a lot of testing with this implementation without a problem showing up, will gladly send packet captures if you want to try. I'm pretty certain it's not a straight sum like your example.

> Sorry for asking all this. You had already done a great job of cracking the codes. But if I change the code after we merge this I won't be able to test the changes because I don't have the device.

It's actually nice to have someone looking over my code as I don't have a lot of experience.